### ISSUE TITLE
Error growth fast in the extrapolation region of t

### ISSUE THREAD

**Issue by smao-astro** [cics-nd/ar-pde-cnn#5]
Hi,

I am very interested in your work and thank you very much make the code public.

From the Figure 11 and 20 in the paper, for deterministic model, the error growth much faster in extrapolation region than in interpolation region of t and I have two questions on it:
1. I learnt that the model is auto-regressive, which means it does not have a sense of `t`, so does that mean that the model, if well trained on all the inputs, should have steadily growing error over t (in both extrapolation and interpolation region)? 
2. I saw the error growth much faster once get into extrapolation region, could you explain why?

Thanks.

**Comment by NickGeneva** [comment]
Hi,

Thanks for your interest is the paper, I appreciate it. Regarding your questions:

1. As you mentioned, because the model is auto-regressive, the error will typically always be expected to accumulate. However, the rate at which the error increases is dependent on two factors: the time range on which you train and the physics of the problem. 
Say you train your model on time-steps `0-200` and then want to predict `0-400`. If the physical phenomena of the system differs between the time range `0-200` and `200-400`, then we should expect the model to have a sharper increase in error in the extrapolation range (potentially even diverging) because it is predicting new physics.
For systems where the physics remain fairly similar between all time-steps (for example the K-S system) then we could expect the model to have stable predictions far into the extrapolation range.
2. Slightly different physics, for the viscous Burgers system (Fig. 11) the first set of time-steps is all about shock interactions, but the later in the extrapolation range the system is all about slow diffusion of the shock wave. While related, the model is trained on the initial shock interaction physics of the system and knows a lot less about how the physics of the diffusion process should play out. Hence you have a non-linear error accumulation, because of the non-linear physics.

Hope this clears things up.

**Comment by smao-astro** [comment]
So may I say that is basically sampling issue and sampling issue does have significant influence on the performance of the model on extrapolation region: if the model under-trained on some inputs (physics in extrapolation region), then it won't perform well. 

So one observation is that the issue here is very similar to training on labeled data: hard to predict accurately when extrapolation, or may I explain it this way: the method is STILL training using labeled data, the highlight point is that: 1. The labeled data is generated on the fly, and since you only generate data needed for next step, it is more efficient. 2. the generation of the labeled data is obeying discretized PDE, so you call it physics-constraint training. (I am not criticizing, just trying to understand the idea of your paper, even though it is still very interesting.)

Ideally I hope to find a way so that the neural network efficiently and directly learn the PDE and solve the PDE, rather than learn from data generated from PDE, I do not know whether this is a naive idea, what do you think?

Thanks.

**Comment by NickGeneva** [comment]
No worries, and yes, the way we pose the training it is similar to basically generating the targets on the fly. However, it is still evaluating the PDE residual at the end of the day, granted a discretized one.  This allows us to use implicit time integration methods to discretize the PDE, but not have to do any costly computations like matrix inverses (ie no numerical solver is needed). So physics-constrained learning can be viewed as assessing the quality of the predictions on the fly during training via the PDE residual. Its a matter of how you frame it I suppose. Of course this implies you know the equation ahead of time.

Regarding learning the solution of a PDE with a known equations this is still the same idea of you have a prediction and evaluate residuals. This would be basically the same methods used all physics constrained work such as [PINNs](https://doi.org/10.1016/j.jcp.2018.10.045), [Deep Ritz](https://arxiv.org/pdf/1710.00211.pdf), or [Zhu et al](https://arxiv.org/pdf/1901.06314.pdf). These ideas, similar to this work are trying to find the solution of the PDE parameterized by a DNN using the residual (or related energy function) of the governing equation as a loss. But you could also learn components of the PDE, or just the RHS of the PDE and use numerical methods to evaluate the time derivative (this is basically how Neural ODEs work).

Now if you don't know the PDE equation but have data that's an entirely different problem where you could used some library based methods such as [SINDY](https://www.pnas.org/content/pnas/113/15/3932.full.pdf), also perhaps a PINNs based approach.

Perhaps what you envision is different than these approaches here.
Hope this helps to some degree.

Cheers.

**Comment by smao-astro** [comment]
Hi @NickGeneva ,

Thank you very much for your explanation!

Yes, I agree with you that this work and PINN are similar on using the governing equations as loss, the difference is that this work is using discretized PDE while PINN is PDE, however from my observation since the loss can not be zero in reality, the model won't exactly obey the governing equation, and thus the prediction is a not so accurate solution of the equation, or an accurate solution of an approximated equation, which prevents the model achieve as high accuracy as numerical method. 

The thought above encouraged me to think whether there is a more direct way to learn the PDE and get as accurate solution at least as direct numerical solver.

**Comment by NickGeneva** [comment]
Great! Best of luck!
